### ISSUE TITLE
Love the site, but it’s WAY over the 10k

### ISSUE THREAD

**Issue by aarongustafson** [timroes/www.timroes.de-10kapart#1]
Adding things up here:

File | GZipped
--------------
atf.min.css | 3.2kB
home.min.css | 2.2kB
main.min.css | 3.0kB
fonts.min.js | 3.1 kB
icons.svg | 6.1kB
filter.svg | 0.5kB
app.min.js | 45.2kKB
favicon192.png | 1kB
https://piwik.werosys.de/piwik.php?idsite=1 | 0.3kB
*TOTAL* | 64kB

I may have missed a few things, but these are in the base experience and are what we are counting toward the 10kB limit. Dropping the `app.min.js` would get you pretty close, then a bit of nip/tuck would help you come in under.

Just an FYI though: when fully loaded, your site is over half a MB.

If you can slim down the base experience to being at or inder the 10kB when GZipped, we can accept it into the contest. For full details on what counts & what doesn’t, see https://a-k-apart.com/faq#size

**Comment by timroes** [comment]
@aarongustafson thanks for your reply.

The sizes you are pointing are gzipped sizes with headers. According to your docs, headers doesn't count? Because e.g. the `atf.min.css` has "only" 2.9kB gzipped.

Also regarding the files, that you track:

* The `app.min.js` is loaded via the `async` attribute, which is supported in all modern browser you stated. Why is this counted against the 10k limit when it's loaded asynchronous? (Same applies for the `fonts.min.js`)
* favicon192.png should actually only be loaded as a favicon, which according to the FAQ doesn't count. Nevertheless the favicon60.png (601 bytes) should be loaded since it's the topimage (but seems not to be in your list).
* `main.min.css` is loaded via the `preload` tag, which is the currently common way, to async load CSS (also see the famous [loadCSS](https://github.com/filamentgroup/loadCSS#recommended-usage-pattern)).
* The `icons.svg` is only used inside the `xlink:href` or use tags. I checked browser behavior for all major browsers, and SVG loaded that way are loaded non-blocking, async, showing the same behavior as if I would have loaded them via async JS myself. So I don't think this should count against the limit.

That piwik is still loaded is strange, I thought I removed that for the contest version, but that didn't seem to work then. Will give it a look, since I am not planning (and think it would anyway not be legal) to track any traffic inside the contest version.

**Comment by aarongustafson** [comment]
The size calculation is here: https://a-k-apart.com/faq#size

Async-ing stuff is great and it page *wait* but doesn’t do anything to reduce page *weight*. If you want to skirt under the limit easily, you could move switch to adding `app.min.js` and `fonts.min.js` via JavaScript. You could maybe lazy load icons.svg via JS too. Lazy loading via JS doesn’t count because if the JS doesn’t run it is never loaded. Async just delays loading.

On a side note, I wonder if concatenating your CSS files might give you an overall smaller CSS footprint when GZip is in play. My gut says yes.

**Comment by timroes** [comment]
I've read the size calculations very carefully before I send the application. Unfortunately several of the points we are discussing here are in my opinion not covered in there and also I never got an answer to the questions I send you. So I assumed common sense (e.g. for the SVG stuff) and tested what I could how browsers behave with this.

Regarding the CSS:  The `main.min.css` and the `atf.min.css` are 5731 bytes when delivered as two files. Concatenating those files would result in 5257 bytes filesize. I don't think, that those 474 bytes saved outweigh the blocking loading of 2.7kB. That's why I prefer only load the critical CSS synchronous, and loading that remaining 2.7kB async. The 1.9kB of the `home.min.css` should clearly stay separate, since I know my visitor statistics and only a very few percentage every visit the home page, so I don't want to bother loading that for all other visitors.

Could you please explain me, in what scenario loading the `app.min.js` and the `fonts.min.js`, by inserting the script tag via JS would benefit anyhow? You say, if the JS never runs, these files will also not be loaded. If JS never runs the browser should (and for what I tested in major browser, they don't) load JS files, so it wouldn't make any difference? Also if you use famous plugins like NoScript to prevent JS execution, I also tested, that the JS files (won't be loaded). Could you please describe a scenario, where loading them via JS improves the usability of the page?

**Comment by aarongustafson** [comment]
Fair point on the CSS, just exploring options for shrinking things down.

In terms of the JS lazy load, I’m thinking of the "broken JS" scenario over
the "JS off" scenario. The GDS findings bear that out: of the 1.1% that did
not get JavaScript enhancements, the vast majority (0.9%) *should* have
gotten the JS enhancement but didn’t for some reason.

https://gds.blog.gov.uk/2013/10/21/how-many-people-are-missing-out-on-javascript-enhancement/

**Comment by timroes** [comment]
Even if we look at the "broken JS" scenario I don't see a use-case. A problem would reside, if the JS files would be loaded, but not executed. I totally agree with you in that, this would be a waste of loaded bytes for nothing. But also looking at the above article, I don't see that scenarios.

Let's have a look at the people with the "broken" JS from the article. Honestly, I thinks the biggest category in this would still be the people using script blockers, that are not mentioned at all in the article. And as I mentioned above those people also don't get the JS transfered.

**Network errors**

For people having network errors, it's no difference in what way the JS loading was triggered. It doesn't matter if it was an `async` script tag or an inline JS triggering those load. So there is no benefit or disadvantage in either technique for them.

**Corporate proxies/blocked scripts**

To block the JS on the network side, you basically have three options:

* Block any requests to JS files. As above, it doesn't matter at all in this scenario, how the JS loading was triggered, it will anyway not be loaded.
* Force the browser not to execute JavaScript (I am sure there are some way through group policies or so). As mentioned in my previous comment, if JS execution is disabled in the browser, the browser shouldn't (and as far as I tested doesn't - and if they do, it's clearly something where the browser should improve) load the JS files. Because that would just be bad browser behaviour to load something, it knows it won't use in any way.
* Removing `script` tags from the HTML. A proxy could possibly remove script tags from HTML. In that case the async loading trigger wouldn't be executed and no further JS would be loaded. But why would a proxy in this case leave the `<script async>` tag inside the HTML, when removing inline JS? I don't think that any proxy would behave the way, that it removes inline scripts, but not external script references.

So I still don't see any way, where loading JS files via `<script async>` does cause the files to load, but not to execute. Still I don't deny there might be some rare cases - and I really would like some heads up on what they could be - where this is an issue, and therefore async loading via JS would be a more reasonable way than using `<script async>`. Let's assume from these 0.85% percentage (calculation from the above article) of unexplained no-JS-execution-visitors, 20% are really running into loading the JS, but not executing it - and I think that number is way to high, since I guess the above mentioned categories cover way more than 80% (and again I can't even think up a scenario, where it would happen). With 20% of 0.85% visitors only 0.17% of all visitors would suffer the problem of loading the JS unnecessarily.

So I don't have a problem introducing changes, that only bring a benefit for 0.17% of all users... as long as it doesn't bring drawbacks for all other users.

Therefore I tested loading the JS via inline script tags instead of the `<script async>` solution. Unfortunately, all browsers I tested had a slightly worse loading experience that way. I assume, because they need to execute JS, that manipulates the DOM, instead of leaving the async loading to the browser, that usually should be able to schedule and handle that the optimized possible way. Also it would increase the size of the HTML slightly due to the async loading JS (which is longer than the word `async`) - but this is really just a very slightly increase in size.

So async loading the JS files via JS, increases the file size and lowers the user experience for 100% of the users, to give 0.17% of users the benefit, of not loading 36 unnecessary kB. Also with the `<script async>` solution, these 0.17% users are not having a broken experience, just some unnecessary transfered kB. I don't think that increasing user experience (and increasing HTML file size, though this is really just by a small part) for 100% of users would outweigh the unnecessary transfered kB for 0.17% of users.

Though I see, that from the pure "progressive enhancement" point of view, the loading via JS would be the better solution, I would totally understand, if you substract 1 or 2 points from the progressive enhancement score, but I don't see how this should prevent this entry from the contest? On the other hand, shouldn't all entries *not* using `<script async>` get one point substracted from the user experience score, for the worse loading behaviour?

**Comment by aarongustafson** [comment]
I appreciate your arguments, but the rules of the contest are the rules of the contest. We’ll discuss, but FWIW every other accepted project has met the threshold of 10k initial load (over 120 of them).
